apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts

```diff
@@ -0,0 +1,1541 @@
+    contentfulRestClientMock,
+  );
+
+  const contentfulGraphqlClientMockServer =
+    getContentfulGraphqlClientMockServer({
+      ResearchOutputs: () => {
+        return getContentfulResearchOutputGraphqlResponse();
+      },
+      ResearchOutputsCollection: () => {
+        return {
+          total: 1,
+          items: [{}],
+        };
+      },
+    });
+
+  const researchOutputDataProviderMockGraphql =
+    new ResearchOutputContentfulDataProvider(
+      contentfulGraphqlClientMockServer,
+      contentfulRestClientMock,
+    );
+
+  beforeAll(() => {
+    jest.useFakeTimers();
+  });
+
+  afterAll(() => {
+    jest.useRealTimers();
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  describe('fetch', () => {
+    test('can call from mock server', async () => {
+      await researchOutputDataProviderMockGraphql.fetch({});
+    });
+
+    test('Should return an empty result when the client returns an empty array of data', async () => {
+      contentfulGraphqlClientMock.request.mockResolvedValueOnce({
+        researchOutputsCollection: {
+          total: 0,
+          items: [],
+        },
+      });
+
+      const result = await researchOutputDataProvider.fetch({
+        take: 10,
+        skip: 5,
+      });
+
+      expect(result).toEqual({ total: 0, items: [] });
+    });
+
+    test('Should return an empty result when the client returns a response with query property set to null', async () => {
+      contentfulGraphqlClientMock.request.mockResolvedValueOnce({
+        researchOutputsCollection: null,
+      });
+
+      const result = await researchOutputDataProvider.fetch({
+        take: 10,
+        skip: 5,
+      });
+
+      expect(result).toEqual({ total: 0, items: [] });
+    });
+
+    describe('Parameters', () => {
+      describe('without search param', () => {
+        beforeEach(() => {
+          contentfulGraphqlClientMock.request.mockResolvedValueOnce({
+            researchOutputsCollection: null,
+          });
+        });
+
+        test('Should pass default pagination parameters if not defined', async () => {
+          await researchOutputDataProvider.fetch({});
+
+          expect(contentfulGraphqlClientMock.request).toHaveBeenCalledWith(
+            expect.anything(),
+            expect.objectContaining({
+              limit: 8,
+              skip: 0,
+            }),
+          );
+        });
+
+        test('Should pass the pagination parameters as expected', async () => {
+          await researchOutputDataProvider.fetch({ take: 13, skip: 7 });
+
+          expect(contentfulGraphqlClientMock.request).toHaveBeenCalledWith(
+            expect.anything(),
+            expect.objectContaining({
+              limit: 13,
+              skip: 7,
+            }),
+          );
+        });
+
+        test('Should pass the parameters to fetch drafts as expected', async () => {
+          await researchOutputDataProvider.fetch({
+            filter: {
```

Comment:
The behaviour is currently undefined because those parameters are [tightly coupled at the router level](https://github.com/yldio/asap-hub/blob/master/apps/crn-server/src/routes/research-output.route.ts#L34).

Having said that, the behaviour of contentful if the draft filter is not set, but the preview API _is_ used is probably problematic (because it will surface unpublished changes to published outputs) so we might want to enforce that same coupling at a data provider level.

I'm not sure if there is any use case where we fetch draft _and_ published outputs in the same request though.

Reply:
So there is a use case (uniqueness checks) and it _is_ problematic.